packages/contracts/contracts/discovery/Batch.sol

```diff
@@ -4,32 +4,277 @@ pragma solidity =0.8.12;
+    uint256 public singleSlotDuration;
+
+    modifier inVotingPeriod() {
+        require(
+            votingPeriod.end >= block.timestamp &&
+                votingPeriod.start <= block.timestamp
+        );
+        _;
+    }
 
     constructor(
         uint256 _batchId,
-        uint256[] memory _projectIds,
+        address[] memory _projectAddresses,
         uint256 _slotCount,
         Period memory _votingPeriod
     ) {
+        controller = msg.sender;
         batchId = _batchId;
-        projectIds = _projectIds;
-        projectCount = _projectIds.length;
+        projectAddresses = _projectAddresses;
         slotCount = _slotCount;
         votingPeriod = _votingPeriod;
+        singleSlotDuration =
+            (_votingPeriod.end - _votingPeriod.start) /
+            _slotCount;
+
+        for (uint256 i = 0; i < projectAddresses.length; i++) {
+            projects[projectAddresses[i]] = Project(
+                projectAddresses[i],
+                ProjectStatus.InProgress
+            );
+        }
+    }
+
+    function vote(
+        address projectAddress,
+        uint256 peoplePoolStake,
+        uint256 stakersPoolStake
+    ) external inVotingPeriod {
+        _vote(projectAddress);
+        _investPeoplePool(peoplePoolStake);
+        _investStakersPool(stakersPoolStake);
+    }
+
+    function getCurrentWinners() external view returns (address[] memory) {
+        return _getWinners();
+    }
+
+    function getProject(address projectAddress)
+        external
+        view
+        returns (Project memory)
+    {
+        address[] memory computedWinners = _getWinners();
+
+        for (uint256 i = 0; i < computedWinners.length; i++) {
+            if (computedWinners[i] == projectAddress) {
+                return Project(projectAddress, ProjectStatus.Won);
+            }
+        }
+
+        if (votingPeriod.end < block.timestamp) {
+            return Project(projectAddress, ProjectStatus.Lost);
+        }
+
+        return projects[projectAddress];
+    }
+
+    function _vote(address projectAddress) internal {
+        require(
+            !userVotesPerProject[projectAddress][msg.sender],
+            "already voted in this project"
+        );
+        _defineWinners();
+        require(
+            projects[projectAddress].status == ProjectStatus.InProgress,
+            "project is not in progress"
+        );
+        userVotesPerProject[projectAddress][msg.sender] = true;
+        userVoteCount[msg.sender]++;
+        projectVoteCount[projectAddress]++;
+
+        if (
+            votes.length == 0 ||
+            (projectToVotesIndex[projectAddress] == 0 &&
+                votesIndexToProject[0] != projectAddress)
+        ) {
+            votes.push(1);
+            votesIndexToProject[votes.length - 1] = projectAddress;
+            projectToVotesIndex[projectAddress] = votes.length - 1;
+        } else {
+            votes[projectToVotesIndex[projectAddress]]++;
+        }
+    }
```

Comment:
I think it would be beneficial if we could extract this logic into a `abstract contract ProjectVoting`, to hide away all of this complicated logic, along with the status & vote count arrays. what do you think? Otherwise I think the Batch contract will be a bit overwhelming soon
Want to pair on this today so I can show you, if you have doubts about how to do it?

Reply:
Let's try it.